Add colors selection to OSC

KikiManjaro opened the pull request:
Related to #1027 

Feature to choose OSC button's background and text colors, to kept text visible at any moment :

<details> 
  <summary>How to change colors</summary>
  
![Screenshot_2022-08-06-17-17-24-360_skyline emu](https://user-images.githubusercontent.com/59285425/183255710-e62dbeea-6ad3-49f7-be71-b32e9ecd5a5d.jpg)
![Screenshot_2022-08-06-17-17-28-029_skyline emu](https://user-images.githubusercontent.com/59285425/183255718-48ee6fe2-8c61-412a-aceb-26c30284d819.jpg)
![Screenshot_2022-08-06-17-17-31-431_skyline emu](https://user-images.githubusercontent.com/59285425/183255720-29c2e3f3-6ac0-4251-8830-4c37fdcb0f8f.jpg)
![Screenshot_2022-08-06-17-17-35-202_skyline emu](https://user-images.githubusercontent.com/59285425/183255722-3818f0bb-6cd2-4ca0-bd05-9e94272abc7d.jpg)

</details>


Here are some exemple of colored layout :

<details> 
  <summary>Exemples</summary>

![Screenshot_2022-08-06-16-57-04-682_skyline emu](https://user-images.githubusercontent.com/59285425/183255754-cfca1a4e-8d80-4882-9426-4913285ebcb1.jpg)
![Screenshot_2022-08-06-16-57-42-419_skyline emu](https://user-images.githubusercontent.com/59285425/183255758-fbdd755a-723f-48bf-aa06-b00cf854a667.jpg)
![Screenshot_2022-08-06-17-43-46-176_skyline emu](https://user-images.githubusercontent.com/59285425/183256260-0cced5f6-26e2-4565-8368-c47f4e018f07.jpg)
![Screenshot_2022-08-06-17-44-24-292_skyline emu](https://user-images.githubusercontent.com/59285425/183256263-13d0fbbf-fa9a-44a7-9955-5f4e809bcfc9.jpg)
![Screenshot_2022-08-06-17-44-52-868_skyline emu](https://user-images.githubusercontent.com/59285425/183256265-c512e242-5fbb-4ade-a859-bfa6a2644e8d.jpg)

</details>

Make OSC text bold and antialiased

<details> 
  <summary>Before</summary>
  
![before](https://user-images.githubusercontent.com/59285425/183312203-fd695d58-0180-4c69-9984-e9437f5481c2.jpg)

</details>

<details> 
  <summary>After</summary>
  
![stick no bold](https://user-images.githubusercontent.com/59285425/183313317-c45494f8-5b6f-40f9-9755-3d62a90eb53b.jpg)

</details>


Fix dpad and button position

<details> 
  <summary>Before</summary>
  
![stick no bold](https://user-images.githubusercontent.com/59285425/183726230-d16e363c-e2c9-4f5f-a747-f71ebeb8f5ba.jpg)

</details>

<details> 
  <summary>After</summary>
  
![dpad and buttons](https://user-images.githubusercontent.com/59285425/183725342-6f7e4248-155b-44fa-8ff8-d19ad1b972c2.jpg)

</details>

KikiManjaro commented:
<details> 
  <summary>Adding non-neon colors</summary>
  
![Screenshot_2022-08-06-17-59-08-077_skyline emu](https://user-images.githubusercontent.com/59285425/183256964-745a5441-e939-4716-b701-f06f6f4806b5.jpg)
![Screenshot_2022-08-06-18-06-39-095_skyline emu](https://user-images.githubusercontent.com/59285425/183256967-a8c3dc47-6778-4972-a5fc-4da09bcce20e.jpg)
![Screenshot_2022-08-06-18-07-49-224_skyline emu](https://user-images.githubusercontent.com/59285425/183256970-d91943ea-4c0e-472a-88d2-d6c3d9ec6402.jpg)

</details>

KikiManjaro commented:
<details> 
  <summary>Change dialog to choose text or bg color to contextMenu</summary>

![skyline_context_menu_osc_colors](https://user-images.githubusercontent.com/59285425/183267893-48010984-6d8f-4f15-9b6a-6330d7a69b98.gif)


</details>

lynxnb commented:
Could you change the color selection dialog to select both background and text color at the same time?

Something like this:
![image](https://user-images.githubusercontent.com/37104290/186665077-53f1f0f6-4dda-4b7e-a036-b16b68553fb0.png)

KikiManjaro commented:
> Could you merge the two color selection dialogs into one that can select both background and text color at the same time?
> 
> Something like this: ![image](https://user-images.githubusercontent.com/37104290/186665077-53f1f0f6-4dda-4b7e-a036-b16b68553fb0.png)

I could try :)

KikiManjaro commented:
d

> Could you merge the two color selection dialogs into one that can select both background and text color at the same time?
> 
> Something like this: ![image](https://user-images.githubusercontent.com/37104290/186665077-53f1f0f6-4dda-4b7e-a036-b16b68553fb0.png)


That's done :) 

<details> 
  <summary>Gif of new merged dialogs</summary>

![Colors Dialog](https://user-images.githubusercontent.com/59285425/186788049-de707d73-7e67-4b46-82a1-6875d4780d2d.gif)


</details>

KikiManjaro commented:
I've added neon purple and neon yellow colors, i've also removed transparent color as it won't fit in the picker and was kind of confusing and not really needed

<details> 
  <summary>Neon Yellow</summary>

![neon yellow](https://user-images.githubusercontent.com/59285425/186990847-8dec3a6a-fe2a-4230-b3fc-f0e638993fe7.jpg)

</details>

<details> 
  <summary>Neon Purple</summary>

![neon purple](https://user-images.githubusercontent.com/59285425/186990860-fad15567-4146-40c6-8f2e-8fb1149fb6b9.jpg)

</details>

<details> 
  <summary>Color Picker</summary>

![palette](https://user-images.githubusercontent.com/59285425/186990904-dddbcf9b-c873-4966-ba42-c9ae2dfa7a7a.jpg)

</details>

KikiManjaro commented:
Fix resetControls for opacity of OSC, mentionned in #1736 by @android-1995

lynxnb commented on app/src/main/java/emu/skyline/input/onscreen/OnScreenButton.kt:
```diff
@@ -90,15 +98,27 @@ abstract class OnScreenButton(
+        when (drawable) {
+            is GradientDrawable -> drawable.setColor(config.backgroundColor)
+            is LayerDrawable -> {
+                for (i in 0 until drawable.numberOfLayers) {
+                    renderColors(drawable.getDrawable(i))
+                }
```
Single statement `for`, drop `{}`.

lynxnb commented on app/src/main/java/emu/skyline/input/onscreen/OnScreenEditActivity.kt:
```diff
@@ -54,35 +61,103 @@ class OnScreenEditActivity : AppCompatActivity() {
+                    binding.onScreenControllerView.setBGColor(colors[position])
+                }
+
+                override fun onCancel() {/*Nothing to do*/
```
Drop the comment here.

lynxnb commented on app/src/main/java/emu/skyline/input/onscreen/OnScreenEditActivity.kt:
```diff
@@ -54,35 +61,103 @@ class OnScreenEditActivity : AppCompatActivity() {
+                    backgroundColorPicker.positiveButton.performClick()
+                }
+
+                override fun onCancel() {/*Nothing to do*/
```
Drop the comment here.

lynxnb commented on build.gradle:
```diff
@@ -11,6 +11,7 @@ buildscript {
     repositories {
         google()
         mavenCentral()
+        jcenter()
```
`JCenter` was deprecated this year, so you need to find the color picker lib on some other repository.

AbandonedCart replied:
If you are married to this one, you can use jitpack
`implementation 'com.github.kristiyanP:colorpicker:v1.1.10'`

lynxnb commented on app/src/main/res/drawable/ic_palette.xml:
```diff
@@ -0,0 +1,5 @@
+<vector android:height="24dp" android:tint="#000000"
+    android:viewportHeight="24" android:viewportWidth="24"
+    android:width="24dp" xmlns:android="http://schemas.android.com/apk/res/android">
+    <path android:fillColor="@android:color/white" android:pathData="M12,3c-4.97,0 -9,4.03 -9,9s4.03,9 9,9c0.83,0 1.5,-0.67 1.5,-1.5 0,-0.39 -0.15,-0.74 -0.39,-1.01 -0.23,-0.26 -0.38,-0.61 -0.38,-0.99 0,-0.83 0.67,-1.5 1.5,-1.5L16,16c2.76,0 5,-2.24 5,-5 0,-4.42 -4.03,-8 -9,-8zM6.5,12c-0.83,0 -1.5,-0.67 -1.5,-1.5S5.67,9 6.5,9 8,9.67 8,10.5 7.33,12 6.5,12zM9.5,8C8.67,8 8,7.33 8,6.5S8.67,5 9.5,5s1.5 ...
```
Run the formatter on this file with `CTRL`+`ALT`+`L`.

lynxnb commented on app/src/main/java/emu/skyline/input/onscreen/OnScreenControllerView.kt:
```diff
@@ -272,6 +273,14 @@ class OnScreenControllerView @JvmOverloads constructor(context : Context, attrs
+        return controls.globalTextColor
+    }
+
+    fun getBGColor() : Int {
```
`getBackgroundColor`, don't shorten to BG. Swap out BG for background also for all other functions in this and in other files.

KikiManjaro replied:
since setBackground is already used by on of the parents i was going with BGColor, i've setted it to BackGround color instead

lynxnb commented on app/src/main/java/emu/skyline/input/onscreen/OnScreenEditActivity.kt:
```diff
@@ -54,35 +61,103 @@ class OnScreenEditActivity : AppCompatActivity() {
         toggleFabVisibility(false)
     }
 
+    private val colors = arrayOf(Color.GRAY, Color.argb(180, 0, 0, 0), Color.argb(180, 255, 255, 255), Color.argb(180, 230, 255, 0), Color.argb(180, 180, 0, 230), Color.argb(180, 255, 60, 40), Color.argb(180, 225, 15, 0), Color.argb(180, 10, 185, 230), Color.argb(180, 70, 85, 245), Color.argb(180, 30, 220, 0))
```
Here you'd want to make an enumeration of all colors with a meaningful name (Black, White, Teal, Lime Green, etc...), then use it to compose the array of colors. Doing it like this makes it very hard to see what colors you have in there without having too look them up.

KikiManjaro replied:
Yep that's done

lynxnb commented on app/src/main/java/emu/skyline/input/onscreen/OnScreenEditActivity.kt:
```diff
@@ -54,35 +61,103 @@ class OnScreenEditActivity : AppCompatActivity() {
+        }
+    }
+
+    private fun mergeColorPickers(mainColorPicker : ColorPicker, secondColorPicker : ColorPicker) {
+        with(mainColorPicker) {
+            dialogViewLayout.apply {
+                layoutParams = FrameLayout.LayoutParams(FrameLayout.LayoutParams.WRAP_CONTENT, FrameLayout.LayoutParams.WRAP_CONTENT).apply { gravity = Gravity.BOTTOM }
+                if (viewTreeObserver.isAlive) {
+                    viewTreeObserver.addOnGlobalLayoutListener(object : OnGlobalLayoutListener {
+                        override fun onGlobalLayout() {
+                            viewTreeObserver.removeOnGlobalLayoutListener(this)
+                            getmDialog()!!.window!!.setLayout((width * 1.2).roundToInt(), (height * 1.8).roundToInt())
+                        }
+                    })
+                }
+            }
+            (secondColorPicker.dialogViewLayout.parent as ViewGroup).removeView(secondColorPicker.dialogViewLayout)
+            getmDialog()?.addContentView(secondColorPicker.dialogViewLayout, LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT))
+        }
+    }
```
You definitely don't want to do this to have both color pickers in the same dialog window. I'm not sure about the library you're using, but you should be able to easily create a new XML layout with both color sections in there, then hook up the code to load up that XML. If you can't do that, you'll need to draw them in code, but there's no need to show both, then dismiss one, then merge them manually like you are doing here.

KikiManjaro replied:
I had to fork the project and create my own color picker with double selector, it was not as easy as i expected but that's done and it works

lynxnb reviewed (changes requested):
Sorry for the very late review, we all forgot about this. I left some comments that you might want to address. If you can't work on this anymore for any reason, let us know.

AbandonedCart commented on app/src/main/java/emu/skyline/input/onscreen/OnScreenButton.kt:
```diff
@@ -90,15 +98,27 @@ abstract class OnScreenButton(
         renderCenteredText(canvas, buttonId.short!!, itemWidth.coerceAtMost(itemHeight) * 0.4f, bounds.centerX().toFloat(), bounds.centerY().toFloat(), alpha)
     }
 
+    fun renderColors(drawable : Drawable) {
```
This could probably be `private`, since it's only referenced locally

KikiManjaro commented:
Sorry didn't have time to makes the changes :) i'll do it asap

KikiManjaro commented:
@lynxnb Sorry for the delay, i've been away for a while but it's better late than never :)

lynxnb commented on app/src/main/java/emu/skyline/input/onscreen/OnScreenConfiguration.kt:
```diff
@@ -22,6 +25,8 @@ interface ControllerConfiguration {
  */
 class ControllerConfigurationDummy(defaultRelativeX : Float, defaultRelativeY : Float) : ControllerConfiguration {
     override var alpha : Int = 255
+    override var textColor = Color.argb(180, 0, 0, 0)
+    override var backgroundColor = Color.argb(180, 255, 255, 255)
```
Use colors from the `Colors` file.

lynxnb commented on app/src/main/java/emu/skyline/input/onscreen/OnScreenConfiguration.kt:
```diff
@@ -32,6 +37,8 @@ class ControllerConfigurationImpl(private val context : Context, private val but
     private inline fun <reified T> config(default : T, prefix : String = "${buttonId.name}_") = sharedPreferences(context, default, prefix, "controller_config")
 
     override var alpha by config(255, "")
+    override var textColor by config(Color.argb(180, 0, 0, 0))
+    override var backgroundColor by config(Color.argb(180, 255, 255, 255))
```
Same here.

lynxnb commented on app/src/main/java/emu/skyline/input/onscreen/OnScreenButton.kt:
```diff
@@ -92,15 +100,25 @@ abstract class OnScreenButton(
     open fun render(canvas : Canvas) {
         val bounds = currentBounds
         val alpha = if (isPressed) (config.alpha - 130).coerceIn(30..255) else config.alpha
+        renderColors(drawable)
```
Since color is only updated by the user, try setting the color to the drawable in an `init` block and provide a way to recreate the buttons when the user changes colors (it might not be needed depending on how `invalidate()` works), rather than setting the color at every render.

lynxnb reviewed (commented):
I've noticed some issues with the color picker dialog, like buttons having a really bright ripple in night mode, and the colors having a slighly darker hexagon shape at their center when in light mode, can you reproduce on your end? I can send screenshots if you need.

I'd also lower the default `alpha` value of the OSC to `155`, as they are too bright with on `255`.
